homeassistant/components/updater.py

```diff
@@ -0,0 +1,143 @@
+
+
+class Updater(Entity):
+    """ Updater entity class """
+
+    hass = None
+    entity_id = 'updater.updater'
+    name = 'Updater'
+
+    def __init__(self, hass, logger, **kwargs):
+        try:
+            from git import Repo
+        except ImportError:
+            logger.error('Missing package gitpython')
+            self._repo_class = None
+        else:
+            self._repo_class = Repo
+
+        self.hass = hass
+        self._logger = logger
+
+        self.config = kwargs
+
+        hass.track_time_change(self.update, hour=0, minute=0, second=0)
+        self.hass.services.register(DOMAIN, 'update', self.run_update)
+
+        self.versions = {'newest_sha': None,
+                         'newest_msg': None,
+                         'newest_date': None,
+                         'newest_url': None,
+                         'current_sha': None}
+
+        self.update()
+
+    def update(self, *args, **kwargs):
+        ''' Update the state of the entity '''
+        _LOGGER.info('Looking for updates.')
+        # pull data from github
+        github_resp = requests.get(
+            GH_API_CALL.format(repo=self.config['repo_name'],
+                               branch=self.config['branch']))
+        github_data = github_resp.json()
+        if github_resp.headers['status'] == '200 OK':
+            self.versions['newest_sha'] = github_data[0]['sha']
+            self.versions['newest_msg'] = github_data[0]['commit']['message']
+            self.versions['newest_date'] = \
+                github_data[0]['commit']['author']['date']
+            self.versions['newest_url'] = github_data[0]['html_url']
+
+        # find local copy info
+        if self._repo_class is not None:
+            repo = self._repo_class(HA_SOURCE_DIR)
+            if repo.bare:
+                self.versions['current_sha'] = None
+            else:
+                self.versions['current_sha'] = repo.head.commit.hexsha
+        else:
+            self.versions['current_sha'] = None
+
+        # update with HA
+        self.update_ha_state()
+
+    @property
+    def hidden(self):
+        ''' only show the entity when an update is available '''
+        return not self.versions['newest_sha'] and \
+            self.versions['current_sha'] == self.versions['newest_sha']
+
+    @property
+    def state(self):
+        ''' Is an update available? '''
+        return not self.hidden
+
+    @property
+    def state_attributes(self):
+        ''' current entity state attributes '''
+        return {'hidden': self.hidden,
+                'local_sha': self.versions['current_sha'],
+                'remote_sha': self.versions['newest_sha'],
+                'message': self.versions['newest_msg'],
+                'date': self.versions['newest_date'],
+                'link': self.versions['newest_url']}
+
+    def run_update(self, *args, **kwargs):
+        ''' run the updater '''
+        upscript = os.path.join(HA_SOURCE_DIR, 'scripts', 'update')
```

Comment:
This method does not depend on anything specific to the updater entity. It would be cleaner if you would extract the registering of the service and this service method to be done inside `setup`.